runtime/onert/api/src/nnfw_api_internal.cc

```diff
@@ -858,26 +873,44 @@ NNFW_STATUS nnfw_session::output_tensorinfo(uint32_t index, nnfw_tensorinfo *ti)
     return NNFW_STATUS_UNEXPECTED_NULL;
   }
 
-  if (index >= primary_subgraph()->getOutputs().size())
-  {
-    std::cerr << "Error during nnfw_session::output_tensorinfo, index is out of range."
-              << std::endl;
-    return NNFW_STATUS_ERROR;
-  }
-
   try
   {
-    auto opidx = primary_subgraph()->getOutputs().at(index);
-    auto shape = primary_subgraph()->operands().at(opidx).shape();
-    // If it is called after `nnfw_run` then get the shape from Execution, not from the graph
-    if (isStateFinishedRun())
+    auto const output_size =
+      (isStateModelLoaded()
+         ? _nnpkg->outputSize()
+         : (_executions.empty() ? _compiler_artifact->_executors->outputSize()
+                                : _executions[0]->primary_parentgraph().getOutputs().size()));
+
+    if (index >= output_size)
+    {
+      std::cerr << "Error during nnfw_session::output_tensorinfo, index is out of range."
+                << std::endl;
+      return NNFW_STATUS_ERROR;
+    }
+
+    if (isStateModelLoaded())
+    {
+      auto info = _nnpkg->outputInfo(index);
+      fillTensorInfo(ti, info.shape(), info.typeInfo().type());
+    }
+    else
     {
-      shape = _execution
-                ? _execution->getOutputShape(onert::ir::IOIndex{index})
-                : _executions.at(_executions.size() - 1)->getOutputShape(onert::ir::IOIndex{index});
+      auto io_index = onert::ir::IOIndex{index};
+      if (_executions.empty())
+      {
+        auto shape = _execution->getOutputShape(io_index);
+        auto dtype = _compiler_artifact->_executors->outputInfo(io_index).typeInfo().type();
+        fillTensorInfo(ti, shape, dtype);
+      }
+      else
+      {
+        auto shape = _executions[0]->getOutputShape(io_index);
+        auto op_index = _executions[0]->primary_parentgraph().getOutputs().at(io_index);
+        auto dtype =
+          _executions[0]->primary_parentgraph().operands().at(op_index).typeInfo().type();
+        fillTensorInfo(ti, shape, dtype);
```

Comment:
We may need abstract executors 

- `IExecutors` interface
  - `SingleMoelExecutors`
  - `MultiModelExecutors`
  - `PipelineExecutors`